portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/site/LessonsTreeView.java

```diff
@@ -0,0 +1,262 @@
+        objectToSerialize.put("pages", pageData);
+        objectToSerialize.put("i18n", getI18n());
+
+        return JSONObject.toJSONString(objectToSerialize);
+    }
+
+    // Return a mapping of PageID -> list of additional items to show
+    private Map<String, List<Map<String, String>>> getAdditionalLessonsPages(final List<String> pageIds) {
+        Connection connection = null;
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+
+        final Map<String, List<Map<String, String>>> result = new HashMap<>();
+
+        try {
+            try {
+                connection = SqlService.borrowConnection();
+                ps = connection.prepareStatement(buildSQL(connection, pageIds));
+
+                ps.setString(1, this.userId);
+
+                for (int i = 0; i < pageIds.size(); i++) {
+                    ps.setString(i + 2, pageIds.get(i));
+                }
+
+                rs = ps.executeQuery();
+
+                while (rs.next()) {
+                    if (!this.isInstructor && hiddenFromStudents(rs)) {
+                        continue;
+                    }
+
+                    if (!result.containsKey(rs.getString("sakaiToolId"))) {
+                        result.put(rs.getString("sakaiToolId"), new ArrayList<Map<String, String>>());
+                    }
+
+                    result.get(rs.getString("sakaiToolId")).add(makeMap(rs));
+                }
+
+            } finally {
+                if (ps != null) { ps.close(); }
+                if (rs != null) { rs.close(); }
+
+                if (connection != null) {
+                    SqlService.returnConnection(connection);
+                }
+            }
+        } catch (SQLException e) {
+            log.error("Failed to get lessons tree: " + e.getMessage());
+            e.printStackTrace();
+        }
+
+        return result;
+    }
+
+    private String buildSQL(final Connection conn, final List<String> pageIds) {
+        return ("SELECT p.toolId as sakaiPageId," +
+            " p.pageId as lessonsPageId," +
+            " s.site_id as sakaiSiteId," +
+            " s.tool_id as sakaiToolId," +
+            " i.id as lessonsItemId," +
+            " i.name as name," +
+            " i.description as description," +
+            " " + toChar(conn, "i.sakaiId") + " as sendingPage," +
+            " p2.hidden as hidden," +
+            " p2.releaseDate as releaseDate," +
+            " log.complete as completed," +
+            " i.required," +
+            " i.prerequisite" +
+            " FROM lesson_builder_pages p" +
+            " INNER JOIN SAKAI_SITE_TOOL s" +
+            "   on p.toolId = s.page_id" +
+            " INNER JOIN lesson_builder_items i" +
+            "   on (i.pageId = p.pageId AND type = 2)" +
+            " INNER JOIN lesson_builder_pages p2" +
+            "   on (p2.pageId = i.sakaiId)" +
+            " LEFT OUTER JOIN lesson_builder_log log" +
```

Comment:
I'm surprised to see lesson_builder_log in this query. I figured this table was just logging. Is this table actually crucial for deciding completion?